packages/block-editor/src/store/selectors.js

```diff
@@ -1540,6 +1540,59 @@ export function getTemplateLock( state, rootClientId ) {
 	return getBlockListSettings( state, rootClientId )?.templateLock ?? false;
 }
 
+/**
+ * Determines if the given block type is visible in the inserter.
+ * Note that this is different than whether a block is allowed to be inserted.
+ * In some cases, the block is not allowed in a given position but
+ * it should still be visible in the inserter to be able to add it
+ * to a different position.
+ *
+ * @param {Object}        state           Editor state.
+ * @param {string|Object} blockNameOrType The block type object, e.g., the response
+ *                                        from the block directory; or a string name of
+ *                                        an installed block type, e.g.' core/paragraph'.
+ *
+ * @return {boolean} Whether the given block type is allowed to be inserted.
+ */
+const isBlockVisibleInTheInserter = ( state, blockNameOrType ) => {
+	let blockType;
+	let blockName;
+	if ( blockNameOrType && 'object' === typeof blockNameOrType ) {
+		blockType = blockNameOrType;
+		blockName = blockNameOrType.name;
+	} else {
+		blockType = getBlockType( blockNameOrType );
+		blockName = blockNameOrType;
+	}
+	if ( ! blockType ) {
+		return false;
+	}
+
+	const { allowedBlockTypes } = getSettings( state );
+
+	const isBlockAllowedInEditor = checkAllowList(
+		allowedBlockTypes,
+		blockName,
+		true
+	);
+	if ( ! isBlockAllowedInEditor ) {
+		return false;
+	}
+
+	// If parent blocks are not visible, child blocks should be hidden too.
+	if ( !! blockType.parent?.length ) {
+		return blockType.parent.some(
```

Comment:
At this point, how do we know `blockType.parent` is an array? Can't it be a string, for example, that also has a `length` property?

For context, I'm looking at [an issue in the support forum](https://wordpress.org/support/topic/error-after-update-228/) that I believe to be related to this check: the report is recent, the users report issues when inserting blocks, and there's no other place in the codebase that uses `parent.some`.

Reply:
Looking at the block registration, it seems `parent` should be null or an array, but we don't enforce that.

Even the [recently added logic](https://github.com/WordPress/gutenberg/pull/66121/files#diff-30ce126b90014d659302087765876898f16562219f796796ed407fdb346c8e4dR235) to block registration doesn't check that parent is an array. This:

```js
if (
  1 === settings?.parent?.length &&
  name === settings.parent[ 0 ]
) { /**/ }
```

would still be fine for blocks that register `parent` as a string.

Reply:
One thing we can do to fix this is to make the if check above explicit for what we want to do: verify that it is an array and that it has elements.

The more general question is: how `parent` can become an array? Are block authors mis-registering this property? Can we now enforce it being an array?

Reply:
PR at https://github.com/WordPress/gutenberg/pull/66234

Reply:
> Even the [recently added logic](https://github.com/WordPress/gutenberg/pull/66121/files#diff-30ce126b90014d659302087765876898f16562219f796796ed407fdb346c8e4dR235) to block registration doesn't check that parent is an array.

Regarding this conditional statement, I suggested a change in [this comment](https://github.com/WordPress/gutenberg/pull/66121#discussion_r1802296536). However, I noticed that, unlike my comment, the order of each conditional was reversed when it was merged, which is certainly not what was intended.

Reply:
cc @gziolo for thoughts about enforcing the `blockType.parent` being an array at registration. Would that be feasible?

Reply:
We are discussing it in this thread: https://github.com/WordPress/gutenberg/pull/66234#discussion_r1806378362. The short answer is if we know that block types were registered with the `parent` set to a string, the best way would be to respect that and provide a fallback handling.